File: lib/finita/module.rb
# frozen_string_literal: true


require 'autoc/module'
require 'autoc/composite'


module Finita

  
  class Module < AutoC::Module

    def initialize(*args, **kws)
      super
      self << @code = Code.new(self)
    end

    def self.render(name, &code)
      m = Module.new(name)
      yield(m) if block_given?
      m.render
    end

    private def total_entities
      @total_entities ||= begin
        (set = super).each do |e|
          @code.setup_entities << e if e.respond_to?(:setup)
          @code.cleanup_entities << e if e.respond_to?(:cleanup)
        end
        set
      end
    end

  end


  class Module::Code

    include AutoC::Entity

    private def setup = @setup ||= "#{@module.name}Setup"
    private def cleanup = @cleanup ||= "#{@module.name}Cleanup"

    attr_reader :setup_entities
    attr_reader :cleanup_entities

    def initialize(m)
      @module  = m
      @setup_entities = []
      @cleanup_entities = []
      dependencies << Module::DEFINITIONS << Module::INCLUDES
    end

    def interface_declarations(stream)
      stream << %{
        /**
          @brief #{@module.name} module initializer
        */
        AUTOC_EXTERN void #{setup}(void);
        /**
          @brief #{@module.name} module finalizer
        */
        AUTOC_EXTERN void #{cleanup}(void);
      }
    end

    def definitions(stream)
      stream << %{
        static int #{@module.name}Live = 0;
      }
      stream << "void #{setup}() { if(!#{@module.name}Live) {"
        @setup_entities.sort.each { |e| e.setup(stream) }
      stream << "#{@module.name}Live = 1;}}"
      stream << "void #{cleanup}() { if(#{@module.name}Live) {"
        @cleanup_entities.sort.reverse_each { |e| e.cleanup(stream) }
      stream << "#{@module.name}Live = 0;}}"
    end

  end


end

File: src/finita/module.py
import sys
import autoc.core
import autoc.module


_context = None


#
class Module(autoc.module.Module):
  
  def __enter__(self):
    self.__context = sys.modules[__name__]._context
    sys.modules[__name__]._context = self
    return super().__enter__()
    
  def __exit__(self, *args):
    x = super().__exit__(*args)
    sys.modules[__name__]._context = self.__context
    return x
  
  
#
class Entity(autoc.module.Entity):
  
    def __init__(self, *args, **kws):
      super().__init__(*args, **kws)
      if _context:
        _context.add(self)
        
        
#
class Variable(autoc.core.Variable, Entity):
  
  def __init__(self, type, name, *args, default=None, **kws):
    t = autoc.core._type(type)
    super().__init__(type, name, *args, dependencies=(t, autoc.core._linkage_code), **kws)
    self.default = default
    
  def render_declarations(self, stream, header):
    super().render_declarations(stream, header)
    if header:
      stream.append(f"AUTOC_EXTERN {self.definition};")

  def render_definitions(self, stream, header):
    super().render_definitions(stream, header)
    if not header:
      stream.append(f"{self.definition}{f" = {self.default}" if self.default else str()};")
      
      
#
def int(name, value=0):
  return Variable("int", name, default=value)


  #
def double(name, value=0):
  return Variable("double", name, default=value)